crates/sparrow-session/src/execution.rs

```diff
@@ -99,29 +95,51 @@ impl Execution {
 
     pub fn next_blocking(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
         self.is_done()?;
```

Comment:
Thought: We can probably cleanup a lot of this if we treat this as "only use the execution once".

1. `is_done()` doesn't really need to do any async stuff. It could just check to see if we've "used" up the future. Then `next_blocking` just needs to make sure it's `Running` (and/or throw an error if we want it to).
2. `collect_all` can just iterate collecting things until the stream is empty, and then it can run down the progress updates to see if it ends in `Ok()` in which case it returns the collected batches or `Err(...)` in which case it returns the error. May be simpler than trying to use the `select`, and handles the race condition noted.